java/oauth-custom-challengehandler/src/main/java/com/esri/arcgisruntime/sample/oauthcustomchallengehandler/MainActivity.java

```diff
@@ -0,0 +1,351 @@
+     * @return the response that should be taken
+     */
+    @Override
+    public AuthenticationChallengeResponse handleChallenge(AuthenticationChallenge challenge) {
+
+      AuthenticationChallengeResponse ret = null;
+
+      if (challenge.getType() == AuthenticationChallenge.Type.OAUTH_CREDENTIAL_CHALLENGE) {
+        final OAuthChallengeCountDownLatch oAuthChallengeCountDownLatch = new OAuthChallengeCountDownLatch(1);
+
+        final String portalUrl = challenge.getRemoteResource().getUri();
+
+        // get the OAuth config for the portalUrl associated with the challenge
+        try {
+          mOAuthConfiguration = AuthenticationManager.getOAuthConfiguration(portalUrl);
+        } catch (MalformedURLException e) {
+          String errorToast = "Exception is thrown when retrieving OAuth configuration. ";
+          String errorLogcat = "Exception is thrown when retrieving OAuth configuration: " + e.getMessage();
+          return handleError(errorToast, errorLogcat, challenge);
+        }
+
+        // Validate OAuth configuration
+        if (mOAuthConfiguration == null) {
+          String error = "OAuth configuration has not been set up yet. ";
+          return handleError(error, error, challenge);
+        }
+        
+        runOnUiThread(new Runnable() {
+          @Override
+          public void run() {
+
+            // get the authorization url which loads the OAuth login page to display in the dialog
+            final String url = OAuthTokenCredentialRequest
+                .getAuthorizationUrl(portalUrl, mOAuthConfiguration.getClientId(), mOAuthConfiguration.getRedirectUri(), 0);
+
+            // setup web view
+            WebView webView = new WebView(MainActivity.this);
+            webView.setLayoutParams(new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
+                ViewGroup.LayoutParams.MATCH_PARENT));
+            WebSettings webSettings = webView.getSettings();
+            webSettings.setBuiltInZoomControls(true);
+            webSettings.setJavaScriptEnabled(true);
+
+            // setup web view client
+            CustomWebViewClient customWebViewClient = new CustomWebViewClient(portalUrl,
+                mOAuthConfiguration.getClientId(), mOAuthConfiguration.getRedirectUri(), oAuthChallengeCountDownLatch);
+            webView.setWebViewClient(customWebViewClient);
+
+            webView.loadUrl(url);
+            // set the webview on the dialog
+            mOAuthLoginDialog.setContentView(webView);
+            // display the dialog
+            mOAuthLoginDialog.show();
+          }
+        });
+
+        try {
+          // wait for the OAuth browser page to be shown and user to login and for the CustomWebViewClient
+          // to get the auth code and request an OAuthTokenCredential with it
+          oAuthChallengeCountDownLatch.await();
+        } catch (InterruptedException ie) {
+          String errorToast = "Operation is interrupted when signing in. ";
```

Comment:
Operation has been interrupted when signing in.

Reply:
updated.